worker/periodicworker_test.go

```diff
@@ -45,6 +46,60 @@ func (s *periodicWorkerSuite) TestWait(c *gc.C) {
 	}
 }
 
+type testNextPeriod struct {
+	jtesting.Stub
+}
+
+func (t *testNextPeriod) nextPeriod(period time.Duration, jitter float64) time.Duration {
+	t.MethodCall(t, "nextPeriod", period, jitter)
+	return period
+}
+
+func (s *periodicWorkerSuite) TestWaitWithJitter(c *gc.C) {
+	funcHasRun := make(chan struct{}, 1)
+	doWork := func(_ <-chan struct{}) error {
+		funcHasRun <- struct{}{}
+		return nil
+	}
+
+	tPeriod := &testNextPeriod{}
+	cleanup := jtesting.PatchValue(&nextPeriod, tPeriod.nextPeriod)
+	defer cleanup()
+
+	w := NewPeriodicWorker(doWork, testing.ShortWait, NewTimer, Jitter(0.2))
+	defer func() { c.Assert(worker.Stop(w), gc.Equals, nil) }()
+
+	select {
+	case <-funcHasRun:
+	case <-time.After(testing.LongWait):
+		c.Fatalf("The doWork function should have been called by now")
+	}
+
+	tPeriod.CheckCalls(c, []jtesting.StubCall{{
+		FuncName: "nextPeriod",
+		Args:     []interface{}{testing.ShortWait, float64(0.2)},
+	}})
+	tPeriod.ResetCalls()
+
+	select {
+	case <-funcHasRun:
+	case <-time.After(testing.LongWait):
+		c.Fatalf("The doWork function should have been called by now")
+	}
+	tPeriod.CheckCalls(c, []jtesting.StubCall{{
+		FuncName: "nextPeriod",
+		Args:     []interface{}{testing.ShortWait, float64(0.2)},
+	}})
+	w.Kill()
+	c.Assert(w.Wait(), gc.Equals, nil)
+
+	select {
+	case <-funcHasRun:
+		c.Fatalf("After the kill we don't expect anymore calls to the function")
+	case <-time.After(defaultFireOnceWait):
+	}
```

Comment:
This is much better. Thanks. A test of the default nextPeriod implementation would be nice too. Maybe just run it a number of times with and without jitter and check that numbers in the correct range come out?